### PR TITLE
Update MulticoreJit

### PR DESCRIPTION
**Pull request by clamp03**
- Enable Generic Methods in MulticoreJit
- Enable NDirect Stub in MulticoreJit

### CODE REVIEW

**Comment by clamp03** [comment]
PR for https://github.com/dotnet/runtime/issues/38981
I used some functions for load and encode in IBC and ZAPPER.
Please review it is correctly implemented. Thank you.

File: src/coreclr/src/vm/ceeload.cpp
Comment (jkotas) [inline src/coreclr/src/vm/ceeload.cpp:1461]: It does not sound right to make MultiCoreJit depend on Preffered Zap Module concept.

File: src/coreclr/clrdefinitions.cmake
Comment (jkotas) [inline src/coreclr/clrdefinitions.cmake:151]: Is there a reason to split this into different feature flags? I think this should be just unconditional part of MultiCoreJIT.

Comment (clamp03) [reply src/coreclr/clrdefinitions.cmake:151]: I used it for test separately. I will fix it. Thank you.

File: src/coreclr/src/vm/ceeload.cpp
Comment (clamp03) [reply src/coreclr/src/vm/ceeload.cpp:1461]: I checked LoaderModule, but it is not proper. I will find another way such as module of caller and etc.
And if you have any good idea, please let me know. Thank you.

Comment (jkotas) [reply src/coreclr/src/vm/ceeload.cpp:1461]: Can the multi-core JIT encoding start with the method's Module (ie `MethodDesc::GetModule()`)?

Comment (clamp03) [reply src/coreclr/src/vm/ceeload.cpp]: I cannot find other way yet. I updated to MethodDesc::GetModule().
I checked performance. It is slower than PreferredZapModule over 5%.

Comment (jkotas) [reply src/coreclr/src/vm/ceeload.cpp]: Why it is slower?

Comment (clamp03) [reply src/coreclr/src/vm/ceeload.cpp]: A generic method is encoded based on the module. So the module should know all types in a generic method.
For example, if List<TizenElement>.Add() List is in System module and TizenElement is a struct, MethodDesc::GetModule will return System module. System module doesn't know TizenElement, so it fails to encode. However, PreferredZapModule may return a module of TizenElement.

Comment (jkotas) [reply src/coreclr/src/vm/ceeload.cpp]: > So the module should know all types in a generic method.

This constraint is specific to ngen signature encoding. Signature encoding for multicore jit trace should not need to be limited by it.
.

Comment (clamp03) [reply src/coreclr/src/vm/ceeload.cpp]: > This constraint is specific to ngen signature encoding. 

Yes, I used ZapSig::EncodeMethod to encode the generic method. It has the constraint.
> Signature encoding for multicore jit trace should not need to be limited by it.

You mean it needs new signature encoding for multicore jit using exising signature builder? Or is there any other encoding function which it can use?

Comment (jkotas) [reply src/coreclr/src/vm/ceeload.cpp]: I think you can reuse most of the encoding logic from `ZapSig::EncodeMethod`, but without depending on the preferred zap module algorithm.

Comment (clamp03) [reply src/coreclr/src/vm/ceeload.cpp]: Thank you. I think that I can make encoding logic that handles value type instantiation (if cannot access from the module of the generic method) in a generic method differently.

**Comment by clamp03** [comment]
I think that failures are not caused by this patch.
Could you please review this PR?
Thank you.

File: src/coreclr/src/inc/corpriv.h
Comment (jkotas) [inline src/coreclr/src/inc/corpriv.h:258]: I think reusing the constants defined in CorElementTypeZapSig for the multicore JIT encoding should be fine.

My concern about using ngen-specific functionality for this multi-core JIT recording was just about the preferred zap module algorithm, reusing the rest is ok.

File: src/coreclr/src/vm/siginfo.cpp
Comment (jkotas) [inline src/coreclr/src/vm/siginfo.cpp:991]: Does this really need a new argument everywhere?

I would expect that you can get what you need by initializating `ZapSig::Context` with your multicore-JIT specific callbacks.

File: src/coreclr/src/inc/corpriv.h
Comment (clamp03) [reply src/coreclr/src/inc/corpriv.h:258]: Thank you. I will update it.

File: src/coreclr/src/vm/siginfo.cpp
Comment (clamp03) [reply src/coreclr/src/vm/siginfo.cpp:991]: Thank you. I will fix it.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (jkotas) [inline src/coreclr/src/vm/multicorejitplayer.cpp:901]: Why do we want to limit it this way?

File: src/coreclr/src/inc/corpriv.h
Comment (jkotas) [inline src/coreclr/src/inc/corpriv.h:221]: Why not reuse `ELEMENT_TYPE_MODULE_ZAPSIG` for this one too?

You should be able to use `ZapSigContext` to interpret the index as appropriate.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (jkotas) [inline src/coreclr/src/vm/multicorejitplayer.cpp:981]: This is very similar to existing `ZapSig::DecodeMethod`. Would it make sense to add `ZapSig::Context` to the existing method to reuse the code?

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp:901]: In our environment, multicorejit tries to load modules of applications that cannot be loaded because profiles are collected in the application termination and multicorejit is performed in pre-initialization stage which is before actual application launching. (And I hope to reuse profiled data in other applications.) If modules cannot be loaded in multicorejit, I found it iscached in FailureCache and it cannot be loaded after. I think it can be removed in the PR. However, I need a similar approach to our environment. If you have a better idea, could you please share it?

File: src/coreclr/src/inc/corpriv.h
Comment (clamp03) [reply src/coreclr/src/inc/corpriv.h:221]: Thank you. I will update.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp:981]: Actually it is very similar to Module::LoadIBCMethodHelper. I will try to reuse ZapSig::DecodeMethod or Module::LoadIBCMethodHelper. Thank you.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (jkotas) [reply src/coreclr/src/vm/multicorejitplayer.cpp:901]: I am wondering whether we should load the assembly here without caching the failure.

Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp]: This is what I tested. I first get profile data when an application terminates. And then a process is launched. It pre-compiles methods using multicorejit before actual application launching.(It is like old zygote.) Multicorejit tries to load all modules in the profile data (It has modules of application as well as system.) However, in this stage, it doesn't know modules of application (It fails and cached.) And then when a user clicks an application, it fails to load module of application because it is cached in FailureCache even if it can be loaded.
If I misunderstand how FailureCache works, please let me know. Thank you

Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp]: Thank you! After this patch is merged, I will create a PR on it.

**Review by kouvel — COMMENTED** [review]
Few comments, I'm not very familiar with the ZapSig stuff, otherwise looks good to me. https://github.com/dotnet/runtime/pull/40948 is merged now, the tests just exercise some code paths. Would be good to add test cases for the new types of methods.

**Comment by clamp03** [comment]
@kouvel Thank you so much. I will fix and update soon.

File: src/coreclr/src/vm/multicorejit.h
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.h]: Fixed typo.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp]: Fixed. I missed to choose correctly. Thank you.

Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp]: Fixed. Thank you.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp]: Removed.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp]: Fixed. Thank you.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp:946]: Yes, right. I added a comment. Thank you.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp]: Fixed. I haven't thought about that. Thank you.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp]: I changed from stack allocation to heap allocation. Thank you.

**Review by kouvel — APPROVED** [review]
Thanks!

**Comment by clamp03** [comment]
@kouvel Thank you so much for your hard work as well as your comments.

**Comment by clamp03** [comment]
@jkotas Could you please review this PR?
And, I have a question. I have merged 15 commits. Can I be a member of .NET platform? Could you let me know how to be a member?
Thank you.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (jkotas) [inline src/coreclr/src/vm/multicorejit.cpp:678]: Do we really need a dedicated recording path for PInvoke methods? Can they be recorded as regular methods and only special cased during replay?

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (jkotas) [inline src/coreclr/src/vm/multicorejitplayer.cpp:1211]: Is the `GetOrCreatePrecode()` call necessary here? I do not look right.

**Comment by jkotas** [comment]
> Can I be a member of .NET platform?

We have a team for .NET runtime contributors from Samsung (https://github.com/orgs/dotnet/teams/samsung/members). I have asked around to figure out how to add you.

Can you take a look at members of the samsung team? It may be a good idea to delete people who are no longer working on .NET from that team.

File: src/coreclr/src/vm/multicorejit.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejit.cpp:678]: It is a good idea. I will update. Thank you.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp:1211]: Not necessary. I want to do as much as possible in multicorejitplayer.
I will fix it. Thank you.

**Comment by clamp03** [comment]
Currently, https://github.com/orgs/dotnet/teams/samsung/members is not available to me.
If I can access the page, I will let you know who are not working on .NET.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (kouvel) [inline src/coreclr/src/vm/multicorejitplayer.cpp:666]: The `else` clause above should be `else if (pMethod->IsNDirect())` instead

**Review by kouvel — APPROVED** [review]

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp:666]: Fixed. Thank you.

**Comment by jkotas** [comment]
You should be member of https://github.com/orgs/dotnet/teams/samsung/members now. Could you please check whether you can see the list?

**Comment by clamp03** [comment]
@jkotas Thank you. I can see the list. I will check the people who are working on .NET or not on Monday.

File: src/coreclr/src/vm/multicorejitplayer.cpp
Comment (kouvel) [reply src/coreclr/src/vm/multicorejitplayer.cpp:666]: Hmm I'm not seeing the fix, I see the commit that fixed it but it didn't seem to have made it into the branch somehow

Comment (clamp03) [reply src/coreclr/src/vm/multicorejitplayer.cpp:666]: Oops. I am sorry. I made a mistake. Thank you.

**Comment by clamp03** [comment]
I think the test failures are not related to this patch.

**Comment by clamp03** [comment]
@jkotas Could you review this PR? Thank you.

**Comment by jkotas** [comment]
@kouvel Do you think that this is good to go, or does this need a second review?

File: src/coreclr/src/vm/zapsig.cpp
Comment (kouvel) [inline src/coreclr/src/vm/zapsig.cpp:252]: I'm seeing this assertion fail because `externalTokens == MulticoreJitTokens` when the generic type parameter is a value type from a different assembly. Repro:

```c#
    private static void Main()
    {
        ProfileOptimization.SetProfileRoot(Environment.CurrentDirectory);
        ProfileOptimization.StartProfile("profile.mcj");

        Thread.Sleep(500);

        CallFoo();

        ProfileOptimization.StartProfile(null);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void CallFoo() => Foo(RegexOptions.IgnoreCase);

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void Foo<T>(T t)
    {
    }
```

I'm not familiar with the zapsig code to evaluate those changes or what might be wrong. It may be good to add some tests involving module-load dependencies as part of the type signature. The rest of the changes look good to me.

**Comment by kouvel** [comment]
@clamp03 it would be helpful to preserve the commits from changes made during the review in the future to make it easier to track the changes. The commits can be squashed on merge.

File: src/coreclr/src/vm/zapsig.cpp
Comment (clamp03) [reply src/coreclr/src/vm/zapsig.cpp]: It is my mistake. Actually, I want to pass if externaltokens are Multicorejittokens.
I updated zapsig encoding for multicorejit.
pfnTokenDefinition is not necessary for multicorejit. I removed. And I added some tests.

**Comment by clamp03** [comment]
@kouvel Sorry, I misunderstood the guide. I thought I should squash manually before the merge.

**Comment by kouvel** [comment]
Hope to clarify the guidance in this PR: https://github.com/dotnet/runtime/pull/43664

**Review by kouvel — APPROVED** [review]
LGTM, thanks! @jkotas I'm not familiar with the zapsig code, if those look ok then I think this is good to go.

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by clamp03** [comment]
Thank you!!!

**Pull request merged**
